src/v/storage/readers_cache.cc

```diff
@@ -297,8 +297,19 @@ void readers_cache::dispose_in_background(entry* e) {
                 delete e; // NOLINT
             });
         });
-    } catch (const ss::gate_closed_exception& e) {
+    } catch (const ss::gate_closed_exception& ex) {
         vlog(stlog.debug, "gate closed while disposing reader");
+        /**
+         * since gate is closed and we failed to call finally on the reader we
+         * add it back to _readers list, it will be gracefully closed in
+         * `readers_cache::close`.
+         * NOTE:
+         * _readers intrusive list is supposed to keep resusable runners but
+         * when gate closed exception is thrwon we are certain that no more
+         * oprations will be executed on the cache so we can reuse the _readers
+         * list to gracefully shutdown readers.
+         */
+        _readers.push_back(*e);
```

Comment:
given that closing and waiting on a gate is used as a synchronization mechanism, what prevents here `_readers` from being updated while it is being processed in `readers_cache::stop` after the gate is closed?

Reply:
I've added an additional comment in the `stop()` method and commit adding check if gate is closed in  `readers_cache::get_reader`. In general it is safe to add reader back to `_readers` here since we wait for all in use readers to be returned back to cache before we dispose entries in `_readers` list